File: src/setupSip.ts
/**
 * SIP Trunk + Dispatch Rule Setup Script
 *
 * Run once to configure LiveKit SIP server with:
 * 1. An inbound SIP trunk (accepts calls from Twilio)
 * 2. A dispatch rule (routes calls to individual rooms, triggering the agent)
 *
 * Usage: npx tsx src/setupSip.ts
 */

import 'dotenv/config';
import { SipClient } from 'livekit-server-sdk';

async function setup() {
  const livekitUrl = process.env['LIVEKIT_URL'] || 'http://localhost:7880';
  const apiKey = process.env['LIVEKIT_API_KEY'] || 'devkey';
  const apiSecret = process.env['LIVEKIT_API_SECRET'] || 'secret';

  const sipClient = new SipClient(livekitUrl, apiKey, apiSecret);

  console.log('[SIP Setup] Creating inbound SIP trunk...');

  const trunk = await sipClient.createSipInboundTrunk(
    'LCA Demo Trunk',  // name
    [],                 // numbers (empty = accept any)
    {
      allowedAddresses: [], // Accept from any IP (local dev)
    },
  );

  console.log(`[SIP Setup] Trunk created: ${trunk.sipTrunkId}`);

  console.log('[SIP Setup] Creating dispatch rule...');

  const rule = await sipClient.createSipDispatchRule({
    type: 'individual' as const,
    roomPrefix: 'call-',
  }, {
    trunkIds: [trunk.sipTrunkId],
  });

  console.log(`[SIP Setup] Dispatch rule created: ${rule.sipDispatchRuleId}`);
  console.log('\n[SIP Setup] Done! Configuration:');
  console.log(`  Trunk ID: ${trunk.sipTrunkId}`);
  console.log(`  Dispatch Rule ID: ${rule.sipDispatchRuleId}`);
  console.log(`  Room prefix: call-`);
  console.log('\nNext steps:');
  console.log('  1. Configure Twilio SIP trunk → point to your LiveKit SIP server (port 5060)');
  console.log('  2. Start the agent: npm run agent');
  console.log('  3. Call your Twilio number');
}

setup().catch(console.error);

File: agent/setup_sip.py
"""
SIP Trunk + Dispatch Rule Setup Script

Run once to configure LiveKit SIP server with:
1. An inbound SIP trunk (accepts calls from Twilio/MicroSIP)
2. A dispatch rule (routes calls to individual rooms, triggering the agent)

Usage: python setup_sip.py
"""

import os
import asyncio
from dotenv import load_dotenv
from livekit import api

load_dotenv()


async def setup():
    livekit_url = os.getenv("LIVEKIT_URL", "http://127.0.0.1:7880")
    api_key = os.getenv("LIVEKIT_API_KEY", "devkey")
    api_secret = os.getenv("LIVEKIT_API_SECRET", "secret")

    lk = api.LiveKitAPI(livekit_url, api_key, api_secret)

    # List existing resources (DO NOT DELETE — other projects may use them)
    print("[SIP Setup] Checking existing SIP config...")
    try:
        existing_rules = await lk.sip.list_dispatch_rule(api.ListSIPDispatchRuleRequest())
        print(f"  Existing dispatch rules: {len(existing_rules.items)}")
    except Exception:
        pass

    print("[SIP Setup] Creating NEW inbound SIP trunk (existing ones untouched)...")
    trunk = await lk.sip.create_inbound_trunk(
        api.CreateSIPInboundTrunkRequest(
            trunk=api.SIPInboundTrunkInfo(
                name="LCA Demo Trunk",
                numbers=[],
                allowed_addresses=["0.0.0.0/0"],
            )
        )
    )
    trunk_id = trunk.sip_trunk_id
    print(f"[SIP Setup] Trunk created: {trunk_id}")

    print("[SIP Setup] Creating dispatch rule with agent_name='lca-agent'...")
    rule = await lk.sip.create_dispatch_rule(
        api.CreateSIPDispatchRuleRequest(
            trunk_ids=[trunk_id],
            rule=api.SIPDispatchRule(
                dispatch_rule_individual=api.SIPDispatchRuleIndividual(
                    room_prefix="call-",
                )
            ),
            room_config=api.RoomConfiguration(
                agents=[
                    api.RoomAgentDispatch(agent_name="lca-agent"),
                ],
            ),
        )
    )
    rule_id = rule.sip_dispatch_rule_id
    print(f"[SIP Setup] Dispatch rule created: {rule_id}")

    print("\n[SIP Setup] Done!")
    print(f"  Trunk ID: {trunk_id}")
    print(f"  Dispatch Rule ID: {rule_id}")
    print(f"  Agent Name: lca-agent")
    print(f"  Room prefix: call-")

    await lk.aclose()


if __name__ == "__main__":
    asyncio.run(setup())
